### ISSUE TITLE
Kubelet build fails with local golang installation

### ISSUE THREAD

**Issue by bradmat** [kubernetes/kubernetes#88764]
**What happened**:
I've been able to successfully install using yum, however I’m now trying to build the Kubelet using a local golang installation but I’ve hit the following error on the first build attempt with ‘make WHAT=cmd/kubelet’
```
# k8s.io/kubernetes/cmd/kubelet
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status
!!! [0302 16:55:20] Call tree:
!!! [0302 16:55:20]  1: /home/brad/go/src/k8s.io/kubernetes/hack/lib/golang.sh:714 kube::golang::build_some_binaries(...)
!!! [0302 16:55:20]  2: /home/brad/go/src/k8s.io/kubernetes/hack/lib/golang.sh:853 kube::golang::build_binaries_for_platform(...)
!!! [0302 16:55:20]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0302 16:55:20] Call tree:
!!! [0302 16:55:20]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0302 16:55:20] Call tree:
!!! [0302 16:55:20]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:93: all] Error 1
```
If I run the same command again however, the error disappears and I get a kubelet binary in the _output/bin directory.

However due to the error on first attempt,  I’m not sure if the build is good which leads onto my follow up question. When does the kubelet.service file get created during the installation process, as I can't run a 'kubeadm join' command to test the binary, and the kubelet binary itself fails when run on its own.

**What you expected to happen**:
Successful Kubelet build.

**How to reproduce it (as minimally and precisely as possible)**:
Just running ‘make WHAT=cmd/kubelet’ on the source detached at the v1.17.3 tag.

**Anything else we need to know?**:
Originally installed the kubectl, kubeadm, and kubelet components through yum which worked successfully, and then removed to try build from source.

Would appreciate any guidance for installing with a local installation on the whole. Have spent a lot of time digging around the source to understand the process and what the yum installer does differently to create things like the service file.

**Environment**:
- Kubernetes version (use `kubectl version`): Haven't built/installed kubectl binary, but attempting from v1.17.3 tag
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): CentOS Linux 7 (Core)
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others: go version go1.13.6 linux/amd64

**Comment by neolit123** [comment]
/remove-sig node cluster-lifecycle
/sig release

please try:
```
# make sure your tree is clean and checkout the v1.17.x tag again
make clean
make all WHAT=cmd/kubelet # "all" means that version information will be added as well
```

**Comment by bradmat** [comment]
I deleted the go/src/ directory created, and started again from go get on k8s.io/kubernetes.

Running git status returns: 
```
# On branch master
nothing to commit, working directory clean
```

git checkout tags/v1.17.3 then returns:
```
Note: checking out 'tags/v1.17.3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at 06ad960... Release commit for Kubernetes v1.17.3
```

Unfortunately, the same error still gets thrown when running make all WHAT=cmd/kubelet:
```
+++ [0304 11:53:21] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
+++ [0304 11:53:30] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
+++ [0304 11:53:37] Building go targets for linux/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
+++ [0304 11:53:50] Building go targets for linux/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
+++ [0304 11:54:03] Building go targets for linux/amd64:
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
+++ [0304 11:54:04] Building go targets for linux/amd64:
    cmd/kubelet
# k8s.io/kubernetes/cmd/kubelet
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status

!!! [0304 11:55:53] Call tree:
!!! [0304 11:55:53]  1: /home/brad/go/src/k8s.io/kubernetes/hack/lib/golang.sh:714 kube::golang::build_some_binaries(...)
!!! [0304 11:55:53]  2: /home/brad/go/src/k8s.io/kubernetes/hack/lib/golang.sh:853 kube::golang::build_binaries_for_platform(...)
!!! [0304 11:55:53]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0304 11:55:53] Call tree:
!!! [0304 11:55:53]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0304 11:55:53] Call tree:
!!! [0304 11:55:53]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:93: all] Error 1
```

**Comment by neolit123** [comment]
odd, i cannot reproduce the issue:
```
$git checkout tags/v1.17.3
Note: checking out 'tags/v1.17.3'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 06ad960bfd0... Release commit for Kubernetes v1.17.3

$make all WHAT=cmd/kubelet
open /home/lubo-it/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/bazel-bin/pkg/generated/openapi/zz_generated.openapi~gopath/src/k8s.io/kubernetes/vendor/k8s.io/kube-scheduler/config/v1alpha2/doc.go: no such file or directory
+++ [0304 15:50:09] Building go targets for linux/amd64:
    cmd/kubelet
```

my go version is:
```
$go version
go version go1.13.6 linux/amd64
```

what is yours?

**Comment by bradmat** [comment]
seems to be exactly the same
```$ go version
go version go1.13.6 linux/amd64```

could it be an issue with gcc versioning?

**Comment by neolit123** [comment]
on my side:
```
gcc version 7.2.0 (Ubuntu 7.2.0-8ubuntu3.2) 
```

to try to obtain more information about the error you can try adding `set -x` in `./hack/lib/golang.sh`

**Comment by bradmat** [comment]
Interesting, there's a difference but I doubt it's that:
```
gcc version 8.3.1 20190311 (Red Hat 8.3.1-3) (GCC)
```

set -x is useful to try, did't think of trying that. Can't however spot much more about it, though it is getting thrown directly after the go install interestingly.

**Comment by neolit123** [comment]
try figuring out which command / line fails exactly.

**Comment by bradmat** [comment]
A quick look through and it spits out the `Coverage is disabled` line of build_some_binaries, so seems to be failing following the go install on line 672 https://github.com/kubernetes/kubernetes/blob/master/hack/lib/golang.sh#L672

**Comment by neolit123** [comment]
> A quick look through and it spits out the Coverage is disabled line of build_some_binaries, so seems to be failing following the go install on line 672 https://github.com/kubernetes/kubernetes/blob/master/hack/lib/golang.sh#L672

this is odd, do you have any idea why `go install` might fail on your setup?

**Comment by bradmat** [comment]
Unfortunately not no, it only seems to be on the cmd/kubelet binary specifically that the go install fails. 

```
go install -gcflags ' -trimpath=/home/brad/go/src/k8s.io/kubernetes' -asmflags -trimpath=/home/brad/go/src/k8s.io/kubernetes -ldflags '-s -w -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.buildDate=2020-03-05T13:58:43Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.buildDate=2020-03-05T13:58:43Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMinor=17+'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMinor=17+'\''' k8s.io/kubernetes/cmd/kubelet
# k8s.io/kubernetes/cmd/kubelet
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status
```

All of the previous binaries have ran through the install successfully, and called place_binaries afterwards.

Even more confusing is perhaps that running it again does then call the place_binaries, but I don't have confidence in the built binary because of the previous error.

```
go install -gcflags ' -trimpath=/home/brad/go/src/k8s.io/kubernetes' -asmflags -trimpath=/home/brad/go/src/k8s.io/kubernetes -ldflags '-s -w -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.buildDate=2020-03-05T14:05:02Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.buildDate=2020-03-05T14:05:02Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMinor=17+'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMinor=17+'\''' k8s.io/kubernetes/cmd/kubelet
+ kube::golang::place_bins
```

Do you think it could it be an issue of the build running tasks in parallel?

**Comment by neolit123** [comment]
> go install -gcflags ' -trimpath=/home/brad/go/src/k8s.io/kubernetes' -asmflags -trimpath=/home/brad/go/src/k8s.io/kubernetes -ldflags '-s -w -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.buildDate=2020-03-05T13:58:43Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.buildDate=2020-03-05T13:58:43Z'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitCommit=06ad960bfd03b39c8310aaf92d1e7c12ce618213'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitTreeState=dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitVersion=v1.17.3-dirty'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMajor=1'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/version.gitMinor=17+'\'' -X '\''k8s.io/kubernetes/vendor/k8s.io/component-base/version.gitMinor=17+'\''' k8s.io/kubernetes/cmd/kubelet

are you running this command outside of the bash script / standalone?
if it fails, this seems like a go / gcc issue and not a k8s issue.

i see mentions of the error here:
https://github.com/golang/go/issues/31336#issuecomment-480874233

does building the minimal example with `import "C"` from here produce errors?
https://github.com/golang/go/issues/31293

> Do you think it could it be an issue of the build running tasks in parallel?

could be.

**Comment by neolit123** [comment]
given this is cgo related:
/sig node

**Comment by bradmat** [comment]
Apologies, that command is pulled from the output when I ran make all WHAT=cmd/kubelet with the set -x

Will try the minimal example shortly, I expect it will build though since the other go installs in the script have built.

**Comment by neolit123** [comment]
> Apologies, that command is pulled from the output when I ran make all WHAT=cmd/kubelet with the set -x

try using the command to build the kubelet from the command line without the k8s bash scripts,

**Comment by bradmat** [comment]
Interesting.. running the install command on its own hits the same issue, following an attempt to find a 'dedent' package.

```
go: finding github.com/mohae/deepcopy v0.0.0-20170603005431-491d3605edfb
go: finding github.com/thecodeteam/goscaleio v0.1.0
go: finding github.com/storageos/go-api v0.0.0-20180912212459-343b3eff91fc
go: finding github.com/lithammer/dedent v1.1.0
# k8s.io/kubernetes/cmd/kubelet
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status
```

**Comment by neolit123** [comment]
so this seems like a go bug.
you can try a newer version and also look for the bug/error in their issue tracker.

**Comment by bradmat** [comment]
Yep, that would seem to be the case. It even seems like it found the 'dedent' package so there's another issue at play somewhere with go.

```
ls ~/go/pkg/mod/github.com/lithammer/dedent@v1.1.0/
dedent.go  dedent_test.go  go.mod  LICENSE  README.md
```

Thanks for your help, I'll try to have a look further into go for a solution.

**Comment by neolit123** [comment]
i will leave it to you if you want to close this ticket.

would be nice to catch more detail about this e.g. does this only happen on `gcc version 8.3.1 20190311 (Red Hat 8.3.1-3) (GCC)` or something related to the distro setup.

**Comment by bradmat** [comment]
I'll close it off since it doesn't seem to be Kubernetes related, but will update with further details if I find any later.

**Comment by neolit123** [comment]
thanks

**Comment by bradmat** [comment]
Just as an update on this issue. I've since tried to swap go version installations, gcc versions, and different versions of this repository.

Specifically I've tried the following versions, and a variety of combinations of them - unfortunately all resulting in the same build error:

```
$ go1.13.8 version
go version go1.13.8 linux/amd64
```
```
$ gcc --version
gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
```
```
$ git status
# HEAD detached at v1.17.0
nothing to commit, working directory clean
```

As a result, I'm less inclined to believe it's a go bug as suspected, and perhaps instead something else in the environment?

**Comment by neolit123** [comment]
> As a result, I'm less inclined to believe it's a go bug as suspected, and perhaps instead something else in the environment?

it seems like something that should be tracked in the go issue tracker.
the bug could exist with multiple go versions and there might not be a fix for it yet.

given you reproduced the bug outside of the k8s build tooling i wouldn't say this is a k8s bug.
